Migration Guide for Flutter web slotted rendering. (#5928)
flutter/website: src/docs/release/breaking-changes/platform-views-using-html-slots-web.md

---
title: Using HTML slots to render platform views in the web

> description: iframes in Flutter web used to reload, because of the way some DOM operations were made. A change in the way Flutter web apps render Platform Views makes them much more stable (preventing iframe reloads, and other problems with video tags or forms potentially losing their state).

Reviewer: Remove "much more". It wasn't stable and now it simply is :)

Author: Hah, yep. Removed!

---

## Summary

Flutter now renders Platform Views using [slot elements][] inside of a single,
app-wide [Shadow Root][]. Slot elements can be added/removed/moved around the
Shadow DOM without affecting the underlying slotted content (which is rendered

> in a constant location)

Reviewer: This discusses an implementation detail, which may be useful in the "Context" section below. In the summary I think the developers would be most interested in what happens to the HTML DOM that _they_ created, so I'd suggest that we talk about how we put their platform views into the root element, instead of injecting it into the Flutter's internal DOM structures.

Author: I moved this to the `context` section.

## Context

> The Flutter framework frequently tweaks its Render Tree to optimize the paint

Reviewer: I'm not sure if "render tree" is worth capitalizing.

Author: Capitalization Of Things, The Bane Of My Existence :P

Author: (lower-cased :P)

operations that are ultimately made per frame. In the web, these Render Tree
changes often result in DOM operations.

> Flutter web used to render its Platform Views (`HtmlElementView` widgets)

Reviewer: I'd probably also not capitalize "platform views"

directly into its corresponding position of the DOM.

Using certain DOM elements as the "target" of some DOM operations causes those
elements to lose their internal state. In practice this means that `iframe` tags
are going to reload, `video` players might restart, or a form being edited might
lose the edits made to it.

This change was made to:

* make Platform Views behave in a stable way in Flutter web.
* unify the way Platform Views were rendered in the web for both rendering

> backends (`html` and `canvaskit`).

Reviewer: And:

* provide a predictable location in the DOM to allow developer to reliably use CSS to style HTML inside platform views, and use other standard DOM API, such as `querySelector`, and `getElementById`.

## Description of change

> A Flutter web app is now rendered inside a common Shadow Root in which slot

Reviewer: I'd not capitalize "shadow root", but I would link to MDN or some other reliable source that explains what it is.

elements are used to represent platform views. The actual content of each
platform view is rendered *outside* of the Shadow Root, as a child of the Shadow

> Root host.

Reviewer: I think we can gzip this sentence a bit: "...is rendered as a sibling of the shadow root"

Reviewer: Or, since we're keeping `<flt-glass-pane>` we could just say "direct child of `<flt-glass-pane>`"

> A Flutter web app is now rendered inside a common Shadow Root in which slot
> elements are used to represent platform views. The actual content of each
> platform view is rendered *outside* of the Shadow Root, as a child of the Shadow
> Root host.

Reviewer 2: I'm suggesting a change that corresponds to Yegor's comment However, he might have intended for that second sentence to be deleted? @yjbanov?

```suggestion
A Flutter web app is now rendered inside a common [shadow root][] in which slot
elements represent platform views. The actual content of each
platform view is rendered *outside* of the shadow root, as a child of the shadow
root host.

[shadow root]: https://www.html5rocks.com/en/tutorials/webcomponents/shadowdom-201/
```

### Before

<!-- skip -->
```html
...

<flt-glass-pane>
  ...
  <div id="platform-view">Contents</div> <!-- canvaskit -->
  <!-- OR -->
  <flt-platform-view>
    #shadow-root
    | <div id="platform-view">Contents</div> <!-- html -->
  </flt-platform-view>
  ...
</flt-glass-pane>

...
```

### After

<!-- skip -->
```html
...

<flt-glass-pane>
  #shadow-root
  | ...
  | <flt-platform-view-slot>
  |   <slot name="platform-view-1" />
  | </flt-platform-view-slot>
  | ...
  <flt-platform-view slot="platform-view-1">
    <div id="platform-view">Contents</div>
  </flt-platform-view>
  ...
</flt-glass-pane>

...
```

> Now, when the framework needs to move DOM nodes around, it'll operate over
> `flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
> content defined in the `flt-platform-view` tags outside of the Shadow Root
> (which never move), thus preventing the reloads.

Reviewer 2: ```suggestion
After this change, when the framework needs to move DOM nodes around,
it operates over`flt-platform-view-slot`s, which only contain the `slot` tag that "projects" the
content defined in the `flt-platform-view` tags outside of the shadow root
(which never move), thus preventing the reloads.
```

> From an app's perspective, this change is transparent, **however** this was
> deemed a _breaking change_ because tests that make assumptions about the DOM of
> a Flutter web app can break.

Reviewer 2: ```suggestion
From an app's perspective, this change is transparent. **However**, this is
considered a _breaking change_ because some tests make assumptions
about how the DOM of a Flutter web app might break.
```

## Migration guide

### Code

> The content returned by the users' Platform View Factory used to be resized and

Reviewer: nit: user's

> The content returned by the users' Platform View Factory used to be resized and

Reviewer: nit: s/resized/sized/

> positioned by the framework, but not anymore. Ensure that your returned

Reviewer: I'd insert another sentence between "but not anymore." and "Ensure...":

...but not anymore. Instead, Flutter sizes and positions `<flt-platform-view>`, which is the parent of the content. Ensure...

Reviewer: I'd also replace "Ensure" with "To ensure". Maybe the developer doesn't want to fill the whole space. They only need to set size to 100% if they want to.

> The content returned by the users' Platform View Factory used to be resized and
> positioned by the framework, but not anymore. Ensure that your returned
> `html.Element` takes the whole space allocated for it by the framework, by
> setting its `style.width` and `style.height` properties to `'100%'`:

Reviewer 2: ```suggestion
Previously, the content returned by [`PlatformViewFactory`][] was resized and
positioned by the framework. Instead, Flutter now sizes and positions
`<flt-platform-view>`, which is the parent of the content.
To ensure the `html.Element` that you return takes the whole space
allocated to it, set its `style.width` and `style.height` properties to `'100%'`:

[`PlatformViewFactory`]: {{site.api}}/javadoc/index.html?io/flutter/plugin/platform/PlatformView.html
```

<!-- skip -->
```dart
ui.platformViewRegistry.registerViewFactory(viewType, (int viewId) {
  final html.Element htmlElement = html.DivElement()
    // ..other props
    ..style.width = '100%'
    ..style.height = '100%';
  // ...
  return htmlElement;
});
```

> Otherwise the engine will print a warning message to the console similar to:

Reviewer 2: ```suggestion
Otherwise, the engine prints a warning message to the console similar to:
```

<!-- skip -->
```bash
Height of Platform View type: [$viewType] may not be set. Defaulting to `height: 100%`.

> Set `style.height` to any appropriate value to stop this message.

Reviewer: Hmm, do we need this? Why can't I use relative layout (flexbox, etc) to fill the space?

Author: If you use any other way of filling the space, you need to set height/width to `auto`, or anything that is not empty. In the engine the only thing we look at is that width/height is set.

Author: I'll add a "But I don't want to set width: 100% in my platform view" section

```

or:

<!-- skip -->
```bash
Width of Platform View type: [$viewType] may not be set. Defaulting to `width: 100%`.

> Set `style.width` to any appropriate value to stop this message.

Reviewer: same here

```

### Tests

> Unfortunately, there are no selectors that let one _pierce_ through the Shadow

Reviewer: s/Unfortunately/By design/ (this is why we use it 😈)

DOM boundary to locate elements underneath it, so code that once peeked inside

> of the `flt-glass-pane`, needs to be made Shadow DOM aware. Using the html

Reviewer: nit: remove comma before "needs"

> Unfortunately, there are no selectors that let one _pierce_ through the Shadow
> DOM boundary to locate elements underneath it, so code that once peeked inside
> of the `flt-glass-pane`, needs to be made Shadow DOM aware. Using the html

Reviewer 2: ```suggestion
By design, there are no selectors that let one _pierce_ through the shadow
DOM boundary to locate elements underneath it, so code that once peeked inside
of the `flt-glass-pane` needs to be made shadow DOM aware. Using the html
```

markup from above:

<!-- skip -->
```dart
// Find #some-element inside the app:
final element = document.querySelector('#some-element');
```

Code after migration:

<!-- skip -->
```dart
final shadowRoot = document.querySelector('flt-glass-pane')?.shadowRoot!;

> final element = shadowRoot.querySelector('#some-element');

Reviewer: Why is this necessary? After the migration `#some-element` will no longer be inside the shadow DOM, only the slot. We don't want people looking into Flutter's private DOM structure. It's unpredictable and unstable.

Author: This is what most of the tests that failed were doing. We can document the fact that now they shouldn't drill inside the shadowDOM, as mentioned above, that's probably better!

Author: Removed this, warned people against looking into the shadowDom of the app.

```

It's probably a good practice to not assume `document` is always going to be the
root element of searches, so a helper like this might be convenient:

<!-- skip -->
```dart
/// Locate elements in the correct root of the application, whether it is
/// `document` or the new `shadowRoot` of `flt-class-pane`.
List<Node> findElements(String selector) {
  final ShadowRoot? shadowRoot = document.querySelector('flt-glass-pane')?.shadowRoot;
  return (shadowRoot != null) ?
    shadowRoot.querySelectorAll(selector):
    document.querySelectorAll(selector);
}
```

(See Relevant PRs below for multiple examples of the "migrations" described
above).

## Timeline

Landed in version: 2.3.0-16.0.pre<br>
In stable release: not yet

## References

> Design Doc:

Reviewer 2: ```suggestion
Design doc:
```
